kernel/src/fs/pipe/named_pipe.rs

```diff
@@ -0,0 +1,267 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{
+    num::Wrapping,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use ostd::sync::WaitQueue;
+
+use super::{PipeReader, PipeWriter};
+use crate::{
+    events::IoEvents,
+    fs::{
+        inode_handle::FileIo,
+        utils::{AccessMode, StatusFlags},
+    },
+    prelude::*,
+    process::signal::{PollHandle, Pollable},
+};
+
+/// A handle for a named pipe that implements `FileIo`.
+///
+/// Once a handle for a `NamedPipe` exists, the corresponding pipe object will
+/// not be dropped.
+struct NamedPipeHandle {
+    inner: Arc<PipeObj>,
+    access_mode: AccessMode,
+}
+
+impl NamedPipeHandle {
+    fn new(inner: Arc<PipeObj>, access_mode: AccessMode) -> Arc<Self> {
+        Arc::new(Self { inner, access_mode })
+    }
+
+    fn try_read(&self, writer: &mut VmWriter) -> Result<usize> {
+        if !self.access_mode.is_readable() {
+            return Err(Error::with_message(Errno::EBADF, "not opened for reading"));
```

Comment:
So why you don't want to use `return_errno_with_message` here?
```suggestion
            return_errno_with_message!(Errno::EBADF, "the file is not opened for reading");
```

---

Meanwhile, I found this should have been checked in `InodeHandle`:
https://github.com/asterinas/asterinas/blob/39a541fdeba53dd90cc85d13dcdd3fee916bb53a/kernel/src/fs/inode_handle/dyn_cap.rs#L87-L89

Therefore, `FileIo` should not check this. (In fact, `Arc<dyn Device>` does not accept the access mode as an argument when opening.)

So maybe we should make this an debug assertion:
```suggestion
    // `InodeHandle` checks the access mode before calling methods in `FileIo`.
    debug_assert!(self.access_mode.is_readable());
```

---

Additionally, the checks in `try_read` and `read` seem to be duplicates for no good reason.